File: experiments/commit_snapshot_crosscheck/receive_pair.py
#!/usr/bin/env python3
"""Receive two pinned snapshots and proceed only on projection agreement.

This receiver imports no SQLite module and accepts no ledger path. Original
contribution under Unknown v0.3. ChatGPT (OpenAI), through Mingli Yuan's
authorized account proxy; not his review or correctness guarantee.
"""
import argparse
import hashlib
import json
from pathlib import Path
import signal
import time

PROFILE = "adva.research.commit-snapshot-pair-receive.v0"
SNAPSHOT_PROFILE = "adva.research.commit-snapshot.v0"
SOURCE_PROFILE = "adva.research.decision-ledger-contention.v0"
MAX_WIRE = 262144
WORK = 0
DELTA = [
    "complete saved two-step prefix independently rechecked",
    "pending third step checked at the same complete endpoint",
    "all four history entries retained without capacity increase",
    "one abstract attempt debit checked without external consumption",
]


def tick(count=1):
    global WORK
    WORK += count
    if WORK > 10000:
        raise ValueError("receiver:work-limit")


def canonical(value):
    return json.dumps(value, sort_keys=True, separators=(",", ":"),
                      ensure_ascii=True, allow_nan=False)


def sha(raw):
    tick()
    return hashlib.sha256(raw).hexdigest()


def parse(raw, label):
    if len(raw) > MAX_WIRE:
        raise ValueError(label + ":wire-byte-limit")

    def unique(pairs):
        out = {}
        for key, value in pairs:
            if key in out:
                raise ValueError(label + ":duplicate-key")
            out[key] = value
        return out

    tick()
    return json.loads(raw, object_pairs_hook=unique,
                      parse_constant=lambda _: (_ for _ in ()).throw(ValueError(label + ":nonfinite")))


def exact_dict(value, fields, label):
    tick()
    if type(value) is not dict or set(value) != set(fields):
        raise ValueError(label + ":fields")


def hex_digest(value, label):
    tick()
    if type(value) is not str or len(value) != 64 or any(c not in "0123456789abcdef" for c in value):
        raise ValueError(label + ":sha256")


def balance(value, label):
    exact_dict(value, ("grant", "spent", "remaining"), label)
    tick(3)
    if not all(type(value[key]) is int for key in value):
        raise ValueError(label + ":integer")
    if not (1 <= value["grant"] <= 4 and 0 <= value["spent"] <= value["grant"]):
        raise ValueError(label + ":range")
    if value["remaining"] != value["grant"] - value["spent"]:
        raise ValueError(label + ":balance")


def read_snapshot(path, pin, label):
    hex_digest(pin, label + ":pin")
    raw = Path(path).read_bytes()
    if sha(raw) != pin:
        raise ValueError(label + ":file-pin")
    snapshot = parse(raw, label)
    if raw != (canonical(snapshot) + "\n").encode("utf-8"):
        raise ValueError(label + ":noncanonical-file")
    exact_dict(snapshot, ("profile", "version", "body", "body_sha256"), label)
    if snapshot["profile"] != SNAPSHOT_PROFILE or snapshot["version"] != 0:
        raise ValueError(label + ":profile-or-version")
    hex_digest(snapshot["body_sha256"], label + ":body")
    body = snapshot["body"]
    if sha(canonical(body).encode("utf-8")) != snapshot["body_sha256"]:
        raise ValueError(label + ":body-digest")
    exact_dict(body, ("source_profile", "source_ledger_sha256", "expected_request", "candidate",
                      "checker_fingerprint", "allowance", "snapshot_state", "transition",
                      "builder_source_sha256"), label + ":body")
    if body["source_profile"] != SOURCE_PROFILE:
        raise ValueError(label + ":source-profile")
    hex_digest(body["source_ledger_sha256"], label + ":source-ledger")
    for name in ("checker_fingerprint", "builder_source_sha256"):
        mapping = body[name]
        if type(mapping) is not dict or not mapping:
            raise ValueError(label + ":" + name)
        for key, value in mapping.items():
            if type(key) is not str or not key:
                raise ValueError(label + ":" + name + ":key")
            hex_digest(value, label + ":" + name)
    return body


def validate_projection(projection, expected, candidate):
    if projection["source_profile"] != SOURCE_PROFILE:
        raise ValueError("projection:source-profile")
    if canonical(projection["expected_request"]) != canonical(expected):
        raise ValueError("projection:expected-binding")
    if canonical(projection["candidate"]) != canonical(candidate):
        raise ValueError("projection:candidate-binding")
    exact_dict(candidate, ("profile", "transition_key", "checkpoint_receipt"), "candidate")
    if candidate["profile"] != SOURCE_PROFILE or type(candidate["transition_key"]) is not str:
        raise ValueError("candidate:profile-or-key")
    allowance = projection["allowance"]
    balance(allowance, "allowance")
    initial = expected.get("allowance") if type(expected) is dict else None
    balance(initial, "initial-allowance")
    state = projection["snapshot_state"]
    stored_result = None
    if state == "empty":
        if projection["transition"] is not None or canonical(allowance) != canonical(initial):
            raise ValueError("empty:transition-or-allowance")
        outcome = "ProvenUncommittedLedger"
    elif state == "committed":
        transition = projection["transition"]
        exact_dict(transition, ("transition_key", "payload", "stored_result"), "transition")
        if transition["transition_key"] != candidate["transition_key"] or canonical(transition["payload"]) != canonical(candidate):
            raise ValueError("transition:binding")
        result = transition["stored_result"]
        exact_dict(result, ("outcome", "reason", "checkpoint_retained", "allowance_after",
                            "semantic_delta"), "result")
        if initial["remaining"] <= 0:
            raise ValueError("committed:initial-allowance")
        after = {"grant": initial["grant"], "spent": initial["spent"] + 1,
                 "remaining": initial["remaining"] - 1}
        if canonical(after) != canonical(allowance) or canonical(result["allowance_after"]) != canonical(after):
            raise ValueError("committed:allowance")
        if result["outcome"] != "AcceptedCheckpointContinuation" or result["reason"] != "prefix-and-declared-next-step-checked":
            raise ValueError("result:scope")
        if canonical(result["semantic_delta"]) != canonical(DELTA):
            raise ValueError("result:delta")
        receipt = candidate["checkpoint_receipt"]
        if type(receipt) is not dict or canonical(result["checkpoint_retained"]) != canonical(receipt.get("checkpoint")):
            raise ValueError("result:checkpoint-binding")
        if canonical(receipt.get("allowance_after")) != canonical(after):
            raise ValueError("candidate:allowance-binding")
        outcome, stored_result = "StoredCommitted", result
    else:
        raise ValueError("projection:snapshot-state")
    return outcome, state, allowance, stored_result


def main(request_path, left_path, right_path):
    global WORK
    WORK = 0
    started = time.perf_counter()
    outcome, reason = "UnknownCommitState", "unstarted"
    state = allowance = stored_result = projection_sha256 = None
    left_sources = right_sources = None
    expected = candidate = None
    signal.signal(signal.SIGALRM, lambda *_: (_ for _ in ()).throw(TimeoutError("wall-limit")))
    signal.setitimer(signal.ITIMER_REAL, 3)
    try:
        request = parse(Path(request_path).read_bytes(), "request")
        exact_dict(request, ("profile", "expected_request", "candidate", "left_snapshot_sha256",
                             "right_snapshot_sha256"), "request")
        if request["profile"] != PROFILE:
            raise ValueError("request:profile")
        expected, candidate = request["expected_request"], request["candidate"]
        left = read_snapshot(left_path, request["left_snapshot_sha256"], "left")
        right = read_snapshot(right_path, request["right_snapshot_sha256"], "right")
        left_sources, right_sources = left["builder_source_sha256"], right["builder_source_sha256"]
        if canonical(left_sources) == canonical(right_sources):
            raise ValueError("builders:provenance-not-distinct")
        left_projection, right_projection = dict(left), dict(right)
        del left_projection["builder_source_sha256"]
        del right_projection["builder_source_sha256"]
        left_bytes = canonical(left_projection).encode("utf-8")
        right_bytes = canonical(right_projection).encode("utf-8")
        if left_bytes != right_bytes:
            raise ValueError("builders:projection-disagreement")
        projection_sha256 = sha(left_bytes)
        outcome, state, allowance, stored_result = validate_projection(
            left_projection, expected, candidate)
        reason = "distinct-builders-agree-on-pinned-ledger-projection"
    except Exception as error:
        reason = type(error).__name__ + ": " + str(error)
    finally:
        signal.setitimer(signal.ITIMER_REAL, 0)
    print(json.dumps({
        "profile": PROFILE,
        "outcome": outcome,
        "reason": reason,
        "expected_request": expected,
        "candidate": candidate,
        "stored_result": stored_result,
        "allowance": allowance,
        "snapshot_state": state,
        "projection_sha256": projection_sha256,
        "left_builder_source_sha256": left_sources,
        "right_builder_source_sha256": right_sources,
        "parent_checked": False,
        "debit_delta": 0,
        "retry_authorized": False,
        "sqlite_opened": False,
        "work_units": WORK,
        "wall_seconds": time.perf_counter() - started,
        "native_authority": False,
        "close_authorized": False,
        "free_authorized": False,
    }, sort_keys=True, ensure_ascii=True, allow_nan=False))


if __name__ == "__main__":
    parser = argparse.ArgumentParser()
    parser.add_argument("--request", required=True)
    parser.add_argument("--left", required=True)
    parser.add_argument("--right", required=True)
    args = parser.parse_args()
    main(args.request, args.left, args.right)

File: experiments/commit_snapshot_node_receiver/calibrated/receive_pair.mjs
#!/usr/bin/env node
// Read two pinned commit snapshots without Python, SQLite, or external packages.
// Project-original contribution under Unknown v0.3. ChatGPT (OpenAI), through
// Mingli Yuan's authorized account proxy; not his review or correctness claim.
import { createHash } from "node:crypto";
import { readFileSync } from "node:fs";
import { performance } from "node:perf_hooks";

const PROFILE = "adva.research.commit-snapshot-pair-receive-node.v1";
const REQUEST_PROFILE = "adva.research.commit-snapshot-pair-receive.v0";
const SNAPSHOT_PROFILE = "adva.research.commit-snapshot.v0";
const SOURCE_PROFILE = "adva.research.decision-ledger-contention.v0";
const MAX_WIRE = 262144;
const MAX_DEPTH = 128;
const MAX_WORK = 15360;
const DELTA = [
  "complete saved two-step prefix independently rechecked",
  "pending third step checked at the same complete endpoint",
  "all four history entries retained without capacity increase",
  "one abstract attempt debit checked without external consumption",
];
let work = 0;
const meter = {value_nodes: 0, checks: 0, decoded: Object.create(null)};

function tick(count = 1, category = "checks") {
  meter[category] += count;
  work += count;
  if (work > MAX_WORK) throw new Error("receiver:work-limit");
}

class StrictJsonParser {
  constructor(text, label) {
    this.text = text;
    this.label = label;
    this.index = 0;
    this.nodes = 0;
    this.keys = 0;
    this.containers = 0;
    this.commas = 0;
  }

  fail(reason) {
    throw new Error(`${this.label}:${reason}@${this.index}`);
  }

  whitespace() {
    while (this.index < this.text.length && " \n\r\t".includes(this.text[this.index])) {
      this.index += 1;
    }
  }

  parse() {
    this.whitespace();
    const value = this.value(0);
    this.whitespace();
    if (this.index !== this.text.length) this.fail("trailing-data");
    return value;
  }

  value(depth) {
    tick(1, "value_nodes");
    this.nodes += 1;
    if (depth > MAX_DEPTH) this.fail("depth-limit");
    this.whitespace();
    const c = this.text[this.index];
    if (c === "{") return this.object(depth + 1);
    if (c === "[") return this.array(depth + 1);
    if (c === "\"") return this.string();
    if (c === "-" || (c >= "0" && c <= "9")) return this.integer();
    for (const [word, value] of [["true", true], ["false", false], ["null", null]]) {
      if (this.text.startsWith(word, this.index)) {
        this.index += word.length;
        return value;
      }
    }
    this.fail("value");
  }

  string() {
    const start = this.index;
    this.index += 1;
    let escaped = false;
    while (this.index < this.text.length) {
      const code = this.text.charCodeAt(this.index);
      const c = this.text[this.index];
      if (!escaped && c === "\"") {
        this.index += 1;
        let value;
        try {
          value = JSON.parse(this.text.slice(start, this.index));
        } catch {
          this.fail("string");
        }
        for (const character of value) {
          if (character.codePointAt(0) > 0x7f) this.fail("non-ascii-string");
        }
        return value;
      }
      if (!escaped && code < 0x20) this.fail("string-control");
      if (!escaped && c === "\\") escaped = true;
      else escaped = false;
      this.index += 1;
    }
    this.fail("unterminated-string");
  }

  integer() {
    const start = this.index;
    if (this.text[this.index] === "-") this.index += 1;
    if (this.text[this.index] === "0") {
      this.index += 1;
      if (/[0-9]/.test(this.text[this.index] ?? "")) this.fail("leading-zero");
    } else {
      if (!/[1-9]/.test(this.text[this.index] ?? "")) this.fail("integer");
      while (/[0-9]/.test(this.text[this.index] ?? "")) this.index += 1;
    }
    if (".eE".includes(this.text[this.index] ?? "")) this.fail("non-integer-number");
    const raw = this.text.slice(start, this.index);
    const value = Number(raw);
    if (!Number.isSafeInteger(value)) this.fail("unsafe-integer");
    return value;
  }

  array(depth) {
    this.containers += 1;
    const result = [];
    this.index += 1;
    this.whitespace();
    if (this.text[this.index] === "]") {
      this.index += 1;
      return result;
    }
    while (true) {
      result.push(this.value(depth));
      this.whitespace();
      if (this.text[this.index] === "]") {
        this.index += 1;
        return result;
      }
      if (this.text[this.index] !== ",") this.fail("array-separator");
      this.commas += 1;
      this.index += 1;
    }
  }

  object(depth) {
    this.containers += 1;
    const result = Object.create(null);
    const seen = new Set();
    this.index += 1;
    this.whitespace();
    if (this.text[this.index] === "}") {
      this.index += 1;
      return result;
    }
    while (true) {
      this.whitespace();
      if (this.text[this.index] !== "\"") this.fail("object-key");
      const key = this.string();
      this.keys += 1;
      if (seen.has(key)) this.fail("duplicate-key");
      seen.add(key);
      this.whitespace();
      if (this.text[this.index] !== ":") this.fail("object-colon");
      this.index += 1;
      result[key] = this.value(depth);
      this.whitespace();
      if (this.text[this.index] === "}") {
        this.index += 1;
        return result;
      }
      if (this.text[this.index] !== ",") this.fail("object-separator");
      this.commas += 1;
      this.index += 1;
    }
  }
}

function decode(buffer, label) {
  if (buffer.length > MAX_WIRE) throw new Error(`${label}:wire-byte-limit`);
  let text;
  try {
    text = new TextDecoder("utf-8", { fatal: true }).decode(buffer);
  } catch {
    throw new Error(`${label}:invalid-utf8`);
  }
  const parser = new StrictJsonParser(text, label);
  const value = parser.parse();
  if (Object.hasOwn(meter.decoded, label)) throw new Error("meter:duplicate-role");
  meter.decoded[label] = {
    bytes: buffer.length, blocks64: Math.ceil(buffer.length / 64),
    value_nodes: parser.nodes, keys: parser.keys,
    lexical_tokens: parser.nodes + 2 * parser.keys + parser.containers + parser.commas,
  };
  return value;
}

function canonical(value) {
  if (value === null) return "null";
  if (value === true) return "true";
  if (value === false) return "false";
  if (typeof value === "number") {
    if (!Number.isSafeInteger(value)) throw new Error("canonical:unsafe-integer");
    return String(value);
  }
  if (typeof value === "string") return JSON.stringify(value);
  if (Array.isArray(value)) return `[${value.map(canonical).join(",")}]`;
  if (typeof value === "object") {
    return `{${Object.keys(value).sort().map((key) => `${JSON.stringify(key)}:${canonical(value[key])}`).join(",")}}`;
  }
  throw new Error("canonical:unsupported-value");
}

function sha(buffer) {
  tick();
  return createHash("sha256").update(buffer).digest("hex");
}

function equal(left, right) {
  return canonical(left) === canonical(right);
}

function exactFields(value, fields, label) {
  tick();
  if (value === null || Array.isArray(value) || typeof value !== "object") {
    throw new Error(`${label}:fields`);
  }
  const actual = Object.keys(value).sort();
  const wanted = [...fields].sort();
  if (!equal(actual, wanted)) throw new Error(`${label}:fields`);
}

function hexDigest(value, label) {
  tick();
  if (typeof value !== "string" || !/^[0-9a-f]{64}$/.test(value)) {
    throw new Error(`${label}:sha256`);
  }
}

function balance(value, label) {
  exactFields(value, ["grant", "spent", "remaining"], label);
  tick(3);
  if (![value.grant, value.spent, value.remaining].every(Number.isSafeInteger)) {
    throw new Error(`${label}:integer`);
  }
  if (!(value.grant >= 1 && value.grant <= 4 && value.spent >= 0 && value.spent <= value.grant)) {
    throw new Error(`${label}:range`);
  }
  if (value.remaining !== value.grant - value.spent) throw new Error(`${label}:balance`);
}

function readSnapshot(path, pin, label) {
  hexDigest(pin, `${label}:pin`);
  const raw = readFileSync(path);
  if (sha(raw) !== pin) throw new Error(`${label}:file-pin`);
  const snapshot = decode(raw, label);
  const expectedRaw = Buffer.from(`${canonical(snapshot)}\n`, "utf8");
  if (!raw.equals(expectedRaw)) throw new Error(`${label}:noncanonical-file`);
  exactFields(snapshot, ["profile", "version", "body", "body_sha256"], label);
  if (snapshot.profile !== SNAPSHOT_PROFILE || snapshot.version !== 0) {
    throw new Error(`${label}:profile-or-version`);
  }
  hexDigest(snapshot.body_sha256, `${label}:body`);
  const body = snapshot.body;
  if (sha(Buffer.from(canonical(body), "utf8")) !== snapshot.body_sha256) {
    throw new Error(`${label}:body-digest`);
  }
  exactFields(body, ["source_profile", "source_ledger_sha256", "expected_request", "candidate",
    "checker_fingerprint", "allowance", "snapshot_state", "transition",
    "builder_source_sha256"], `${label}:body`);
  if (body.source_profile !== SOURCE_PROFILE) throw new Error(`${label}:source-profile`);
  hexDigest(body.source_ledger_sha256, `${label}:source-ledger`);
  for (const name of ["checker_fingerprint", "builder_source_sha256"]) {
    const mapping = body[name];
    if (mapping === null || Array.isArray(mapping) || typeof mapping !== "object" || Object.keys(mapping).length === 0) {
      throw new Error(`${label}:${name}`);
    }
    for (const [key, digest] of Object.entries(mapping)) {
      if (!key) throw new Error(`${label}:${name}:key`);
      hexDigest(digest, `${label}:${name}`);
    }
  }
  return body;
}

function projection(body) {
  const result = Object.create(null);
  for (const key of Object.keys(body)) {
    if (key !== "builder_source_sha256") result[key] = body[key];
  }
  return result;
}

function validateProjection(body, expected, candidate) {
  if (body.source_profile !== SOURCE_PROFILE) throw new Error("projection:source-profile");
  if (!equal(body.expected_request, expected)) throw new Error("projection:expected-binding");
  if (!equal(body.candidate, candidate)) throw new Error("projection:candidate-binding");
  exactFields(candidate, ["profile", "transition_key", "checkpoint_receipt"], "candidate");
  if (candidate.profile !== SOURCE_PROFILE || typeof candidate.transition_key !== "string") {
    throw new Error("candidate:profile-or-key");
  }
  const allowance = body.allowance;
  balance(allowance, "allowance");
  const initial = expected !== null && typeof expected === "object" && !Array.isArray(expected)
    ? expected.allowance : null;
  balance(initial, "initial-allowance");
  if (body.snapshot_state === "empty") {
    if (body.transition !== null || !equal(allowance, initial)) {
      throw new Error("empty:transition-or-allowance");
    }
    return { outcome: "ProvenUncommittedLedger", state: "empty", allowance, storedResult: null };
  }
  if (body.snapshot_state !== "committed") throw new Error("projection:snapshot-state");
  const transition = body.transition;
  exactFields(transition, ["transition_key", "payload", "stored_result"], "transition");
  if (transition.transition_key !== candidate.transition_key || !equal(transition.payload, candidate)) {
    throw new Error("transition:binding");
  }
  const result = transition.stored_result;
  exactFields(result, ["outcome", "reason", "checkpoint_retained", "allowance_after", "semantic_delta"], "result");
  if (initial.remaining <= 0) throw new Error("committed:initial-allowance");
  const after = Object.assign(Object.create(null), {
    grant: initial.grant,
    spent: initial.spent + 1,
    remaining: initial.remaining - 1,
  });
  if (!equal(after, allowance) || !equal(result.allowance_after, after)) {
    throw new Error("committed:allowance");
  }
  if (result.outcome !== "AcceptedCheckpointContinuation" || result.reason !== "prefix-and-declared-next-step-checked") {
    throw new Error("result:scope");
  }
  if (!equal(result.semantic_delta, DELTA)) throw new Error("result:delta");
  const receipt = candidate.checkpoint_receipt;
  if (receipt === null || typeof receipt !== "object" || Array.isArray(receipt)
      || !equal(result.checkpoint_retained, receipt.checkpoint)) {
    throw new Error("result:checkpoint-binding");
  }
  if (!equal(receipt.allowance_after, after)) throw new Error("candidate:allowance-binding");
  return { outcome: "StoredCommitted", state: "committed", allowance, storedResult: result };
}

function argumentsOf(argv) {
  const result = Object.create(null);
  for (let i = 0; i < argv.length; i += 2) {
    if (!argv[i]?.startsWith("--") || argv[i + 1] === undefined) throw new Error("arguments:shape");
    result[argv[i].slice(2)] = argv[i + 1];
  }
  for (const key of ["request", "left", "right"]) if (!result[key]) throw new Error(`arguments:${key}`);
  return result;
}

const started = performance.now();
let outcome = "UnknownCommitState";
let reason = "unstarted";
let state = null;
let allowance = null;
let storedResult = null;
let projectionSha256 = null;
let leftSources = null;
let rightSources = null;
let expected = null;
let candidate = null;
try {
  const args = argumentsOf(process.argv.slice(2));
  const request = decode(readFileSync(args.request), "request");
  exactFields(request, ["profile", "expected_request", "candidate", "left_snapshot_sha256",
    "right_snapshot_sha256"], "request");
  if (request.profile !== REQUEST_PROFILE) throw new Error("request:profile");
  expected = request.expected_request;
  candidate = request.candidate;
  const left = readSnapshot(args.left, request.left_snapshot_sha256, "left");
  const right = readSnapshot(args.right, request.right_snapshot_sha256, "right");
  leftSources = left.builder_source_sha256;
  rightSources = right.builder_source_sha256;
  if (equal(leftSources, rightSources)) throw new Error("builders:provenance-not-distinct");
  const leftProjection = projection(left);
  const rightProjection = projection(right);
  const leftBytes = Buffer.from(canonical(leftProjection), "utf8");
  const rightBytes = Buffer.from(canonical(rightProjection), "utf8");
  if (!leftBytes.equals(rightBytes)) throw new Error("builders:projection-disagreement");
  projectionSha256 = sha(leftBytes);
  const checked = validateProjection(leftProjection, expected, candidate);
  outcome = checked.outcome;
  state = checked.state;
  allowance = checked.allowance;
  storedResult = checked.storedResult;
  reason = "distinct-builders-agree-on-pinned-ledger-projection";
} catch (error) {
  reason = `${error?.name ?? "Error"}: ${error?.message ?? String(error)}`;
}

const report = Object.assign(Object.create(null), {
  profile: PROFILE,
  outcome,
  reason,
  expected_request: expected,
  candidate,
  stored_result: storedResult,
  allowance,
  snapshot_state: state,
  projection_sha256: projectionSha256,
  left_builder_source_sha256: leftSources,
  right_builder_source_sha256: rightSources,
  parent_checked: false,
  debit_delta: 0,
  retry_authorized: false,
  sqlite_opened: false,
  ledger_path_accepted: false,
  work_units: work,
  meter,
  wall_nanoseconds: Math.round((performance.now() - started) * 1000000),
  process_peak_rss_kib: process.resourceUsage().maxRSS,
  runtime: `Node.js ${process.version}`,
  native_authority: false,
  close_authorized: false,
  free_authorized: false,
});
process.stdout.write(`${canonical(report)}\n`);
